pkg/catalog/uncommitted_iterator.go

```diff
@@ -0,0 +1,133 @@
+package catalog
+
+import (
+	"context"
+
+	"github.com/treeverse/lakefs/pkg/graveler"
+)
+
+type UncommittedIterator struct {
+	store     Store
+	ctx       context.Context
+	err       error
+	branchItr graveler.BranchIterator
+	entryItr  *valueEntryIterator
+	entry     *UncommittedRecord
+}
+
+type UncommittedRecord struct {
+	branchID graveler.BranchID
+	*EntryRecord
+}
+
+func NewUncommittedIterator(ctx context.Context, store Store, repository *graveler.RepositoryRecord) (*UncommittedIterator, error) {
+	bItr, err := store.ListBranches(ctx, repository)
+	if err != nil {
+		return nil, err
+	}
+	return &UncommittedIterator{
+		store:     store,
+		ctx:       ctx,
+		branchItr: bItr,
+	}, nil
+}
+
+// nextStaging reads the next branch staging area into entryItr
+func (u *UncommittedIterator) nextStaging() bool {
+	if u.entryItr != nil {
+		u.entryItr.Close()
+	}
+	vItr, err := u.store.ListStaging(u.ctx, u.branchItr.Value().Branch)
+	if err != nil {
+		u.err = err
+		return false
+	}
+	u.entryItr = NewValueToEntryIterator(vItr)
+	return true
+}
+
+// next Sets iterators to provide the next entry. Handles dependency between branch and value iterators.
+// Sets value and returns true if next entry available - false otherwise
+func (u *UncommittedIterator) next() bool {
+	u.entry = nil // will stay nil as long as no new value found
+	for u.entry == nil {
+		if u.branchItr.Next() && u.nextStaging() {
+			if u.entryItr.Next() {
+				u.entry = &UncommittedRecord{
+					branchID: u.branchItr.Value().BranchID,
+				}
+				u.entry.EntryRecord = u.entryItr.Value()
+				return true
+			}
+			u.entryItr.Close()
+			if u.entryItr.Err() != nil {
+				return false
+			}
```

Comment:
The "protocol" state that we need to check the error first - as closing the iterator doesn't require to keep the error.